### DISCUSSION TITLE
await for `revalidatePath` and `revalidateTag` in appRouter

### DISCUSSION THREAD

**Discussion by asnaeb** [vercel/next.js#55322] [Ideas]
### Goals

It would be useful if `revalidatePath` and `revalidateTag` from `"next/cache"`allowed to wait for the revalidation to complete, returning `Promise<void>` instead of `void`.

### Background

There is no other way that I am aware of to achieve the same result. 

### Proposal

Example use case using server actions for simplicity, the same concept can be applied to route handlers
```js
// actions.js

"use server"

export async function action(id, pathname) {
  await db.deactivateItem(id);
  await revalidatePath(pathname); // <-- Allows to return from the function only after revalidation happens
}
```

```js
// client.js

"use client"

import {useTransition} from "react";
import {usePathname} from "next/navigation";
import {action} from "./actions.js"

export default function Items({items}) {
  const listItems = items.map(item =>  <Item item={item}/>);
  return <ul>{listItems}</ul>;
}

function Item({item}) {
  const pathname = usePathname();
  const [isLoading, startTransition] = useTransition();
  const handler = () => {
    startTransition(() => {
      action(item.id, pathname).catch();
    });
  };
  let className = "default";
  if (isLoading) {
    className = "loading";
  }
  else if (!item.active) {
    className = "inactive";
  }
  return <li>
    <span className={className}>
      {item.name}
    </span>
    <button onClick={handler}>deactivate</button>
  </li>;
}
```

```js
// page.js

import Items from "./client.js";

export default async function ItemsPage() {
  const items = await db.getItems();
  return <Items items={items}/>
}
```

This way, `"loading"` class will only be removed after revalidation. The same example without `revalidatePath` being a promise, as it is right now, would result in `"loading"` class removed, revalidation happening and `"inactive"` class applied after revalidation, flashing the `"default"` class during `revalidatePath` call.

**Comment by nicholsss** [comment]
U found something to mimic this behavior?

**Comment by maseranw** [comment]
This would be nice, I'm also looking for a way to do this, if it's even possible.
